[RISC-V] Use correct registers set for fast tail call when function needs GSCookie check

yurai007 opened the pull request:
Consider scenario when JIT needs to emit GS Cookie check before performing fast tail call through t0 at the end of function epilog. In that case CodeGen::genEmitGSCookieCheck may blindly override t0 register to GSCookie address without respecting value already stored in t0. We end up with hard to debug crash since our stack contains only one 0x9abcdef012345678 frame.
To prevent such issue, in this patch we follow ARM64 behaviour and don't pick REG_GSCOOKIE_TMP_0/1 registers for fast tail calls. It fixes couple of CoreCLR tests running in JitStress mode.

Part of https://github.com/dotnet/runtime/issues/84834
cc @wscho77 @HJLeee @clamp03 @JongHeonChoi @t-mustafin @gbalykov @ashaurtaev @sirntar @tomeksowi @Bajtazar @viewizard

yurai007 committed:
[RISC-V] Do not perform fast tail call when function need GSCookie check

Consider scenario when JIT need to emit GS Cookie check before performing
fast tail call through t0 at the end of function epilog. In that case CodeGen::genEmitGSCookieCheck
may blindly override t0 register to GSCookie address without respecting value already stored in t0.
We end up with hard to debug crash since our stack contains only one 0x9abcdef012345678 frame.
To prevent such issue, in this patch we copy x64 and ARM behaviour and use tailcall-via-help
instead fast call method. It fixes couple of CoreCLR tests running in JitStress mode.

yurai007 committed:
Merge branch 'main' into fix-jitstress

jakobbotsch commented:
I think you should handle this in the same way as arm64:

https://github.com/dotnet/runtime/blob/754a554c0072ccb2013bbf7e0952ded8a44b56f4/src/coreclr/jit/lsraarmarch.cpp#L168-L180

The existing reason we reject tailcalls when GS cookies are needed on x64 is just conservative. It could be removed.
For ARM32, there just aren't enough volatile registers available in this case, so we have to reject it.

For RISCV, there should be plenty of volatile registers available, so you can just teach LSRA to make sure it picks a register that won't conflict, in the same way as ARM64.

yurai007 committed:
[RISC-V] Use correct registers set for fast tail call when function needs GSCookie check

Consider scenario when JIT needs to emit GS Cookie check before performing
fast tail call through t0 at the end of function epilog. In that case CodeGen::genEmitGSCookieCheck
may blindly override t0 register to GSCookie address without respecting value already stored in t0.
We end up with hard to debug crash since our stack contains only one 0x9abcdef012345678 frame.
To prevent such issue, in this patch we follow ARM64 behaviour
and don't pick REG_GSCOOKIE_TMP_0/1 registers for fast tail calls.
It fixes couple of CoreCLR tests running in JitStress mode.
